### ISSUE TITLE
No address found for accounts.google.com

### ISSUE THREAD

**Issue by bgood** [GoogleCloudPlatform/google-cloud-ruby#1840]
We are trying use the gem inside Google Kubernetes Engine, however, when authenticate using a service account key file this exception is thrown.

```Hostname not known: accounts.google.com at ["/usr/local/lib/ruby/2.1.0/resolv-replace.rb:12:in `rescue in getaddress'", "/usr/local/lib/ruby/2.1.0/resolv-replace.rb:9:in `getaddress'", "/usr/local/lib/ruby/2.1.0/resolv-replace.rb:23:in `initialize'", "/usr/local/lib/ruby/2.1.0/net/http.rb:879:in `open'", "/usr/local/lib/ruby/2.1.0/net/http.rb:879:in `block in connect'", "/usr/local/lib/ruby/2.1.0/timeout.rb:76:in `timeout'", "/usr/local/lib/ruby/2.1.0/net/http.rb:878:in `connect'", "/usr/local/lib/ruby/2.1.0/net/http.rb:863:in `do_start'", "/usr/local/lib/ruby/2.1.0/net/http.rb:852:in `start'", "/usr/local/lib/ruby/2.1.0/net/http.rb:1369:in `request'", "/usr/local/bundle/gems/faraday-0.12.2/lib/faraday/adapter/net_http.rb:80:in `perform_request'", "/usr/local/bundle/gems/faraday-0.12.2/lib/faraday/adapter/net_http.rb:38:in `block in call'"]```

After poking around it seems like if a trailing . is included on accounts.google.com, the resolution works correctly.  Here are working and not working examples from irb.

Failing:
```irb(main):001:0> require 'resolv'
=> true
irb(main):002:0>  p Resolv.getaddress('accounts.google.com')
Resolv::ResolvError: no address for accounts.google.com
	from /usr/local/lib/ruby/2.1.0/resolv.rb:93:in `getaddress'
	from /usr/local/lib/ruby/2.1.0/resolv.rb:43:in `getaddress'
	from (irb):2
	from /usr/local/bin/irb:11:in `<main>'
```

Working:
```irb(main):001:0> require 'resolv'
=> true
irb(main):002:0> p Resolv.getaddress('accounts.google.com.')
"74.125.69.84"
=> "74.125.69.84"
```

**Comment by blowmage** [comment]
@bgood Thanks for opening this issue. Which google-cloud-* gems are you using? Can you also provide the versions? Is there more of the backtrace you can share?

**Comment by blowmage** [comment]
@geigerj Thoughts on this?

**Comment by bgood** [comment]
We are using **google-cloud-pubsub (0.27.2)** among others, but the error is getting thrown when we try to init that gem.  Unfortunately, that is all the trace I have.

Honestly, I'm not sure is this is more of a core Ruby issue, Kubernetes/GKE issue or gem issue, but I figured it is fixable or at least the easiest to workaround in the gem.

**Comment by geigerj** [comment]
That behavior with `Resolv.getAddresss` is perplexing. I'm not able to reproduce in Ruby 2.1.0 outside of GKE, which makes me suspect a GKE issue rather than core Ruby, but I'm not sure.

Since adding the trailing dot shouldn't hurt, I don't have a problem adding it as a workaround here. If you have a stack trace, it might be helpful to determine whether this string is coming from `google-cloud-core`, or from [the `googleauth` library](https://github.com/google/google-auth-library-ruby), which is in a separate repo.

**Comment by bgood** [comment]
Yes, this is an interesting problem.  Even on GKE but outside of Ruby the address resolves, so it doesn't seem to be squarely in any particular technology or system.  Thank you for your willingness to implement a workaround.

Here is more stack trace:

```Hostname not known: accounts.google.com at ["/usr/local/lib/ruby/2.1.0/resolv-replace.rb:12:in `rescue in getaddress'", "/usr/local/lib/ruby/2.1.0/resolv-replace.rb:9:in `getaddress'", "/usr/local/lib/ruby/2.1.0/resolv-replace.rb:23:in `initialize'", "/usr/local/lib/ruby/2.1.0/net/http.rb:879:in `open'", "/usr/local/lib/ruby/2.1.0/net/http.rb:879:in `block in connect'", "/usr/local/lib/ruby/2.1.0/timeout.rb:76:in `timeout'", "/usr/local/lib/ruby/2.1.0/net/http.rb:878:in `connect'", "/usr/local/lib/ruby/2.1.0/net/http.rb:863:in `do_start'", "/usr/local/lib/ruby/2.1.0/net/http.rb:852:in `start'", "/usr/local/lib/ruby/2.1.0/net/http.rb:1369:in `request'", "/usr/local/bundle/gems/faraday-0.12.2/lib/faraday/adapter/net_http.rb:80:in `perform_request'", "/usr/local/bundle/gems/faraday-0.12.2/lib/faraday/adapter/net_http.rb:38:in `block in call'", "/usr/local/bundle/gems/faraday-0.12.2/lib/faraday/adapter/net_http.rb:85:in `with_net_http_connection'", "/usr/local/bundle/gems/faraday-0.12.2/lib/faraday/adapter/net_http.rb:33:in `call'", "/usr/local/bundle/gems/faraday-0.12.2/lib/faraday/request/url_encoded.rb:15:in `call'", "/usr/local/bundle/gems/faraday-0.12.2/lib/faraday/rack_builder.rb:141:in `build_response'", "/usr/local/bundle/gems/faraday-0.12.2/lib/faraday/connection.rb:386:in `run_request'", "/usr/local/bundle/gems/faraday-0.12.2/lib/faraday/connection.rb:186:in `post'", "/usr/local/bundle/gems/signet-0.8.1/lib/signet/oauth_2/client.rb:967:in `fetch_access_token'", "/usr/local/bundle/gems/signet-0.8.1/lib/signet/oauth_2/client.rb:1005:in `fetch_access_token!'", "/usr/local/bundle/gems/googleauth-0.5.3/lib/googleauth/signet.rb:69:in `fetch_access_token!'", "/usr/local/bundle/gems/google-cloud-core-1.0.0/lib/google/cloud/credentials.rb:59:in `initialize'", "/usr/local/bundle/gems/google-cloud-pubsub-0.27.2/lib/google/cloud/pubsub.rb:549:in `new'", "/usr/local/bundle/gems/google-cloud-pubsub-0.27.2/lib/google/cloud/pubsub.rb:549:in `credentials_with_scope'", "/usr/local/bundle/gems/google-cloud-pubsub-0.27.2/lib/google/cloud/pubsub.rb:535:in `new'", "/promotion-management/lib/gcp/consumer.rb:9:in `initialize'", "/promotion-management/lib/tasks/pubsub.rake:9:in `new'", "/promotion-management/lib/tasks/pubsub.rake:9:in `block (2 levels) in <top (required)>'", "/usr/local/lib/ruby/2.1.0/rake/task.rb:234:in `call'", "/usr/local/lib/ruby/2.1.0/rake/task.rb:234:in `block in execute'"]```

**Comment by quartzmo** [comment]
Just for the sake of being up-to-date (I don't know that it will fix the problem), would it be much trouble for you to reproduce the issue with the latest google-cloud-pubsub `0.28.0`?

**Comment by quartzmo** [comment]
Sorry, and google-cloud-core `1.1.0` as well.

**Comment by bgood** [comment]
Looks like this was a bug in Ruby (fixed in 2.1.6 and higher) after all.  We upgraded to the current version of Ruby 2.1 and it is now working.  

For reference purposes here is the link to the Ruby bug (https://bugs.ruby-lang.org/issues/10412)

Sorry for the trouble.

**Comment by quartzmo** [comment]
@bgood No problem, we're always happy to try to help. Thanks for linking to the Ruby issue.
